cooked-validators/src/Cooked/Tx/Constraints/Type.hs

```diff
@@ -53,20 +54,36 @@ type SpendsConstrs a =
     Typeable a
   )
 
--- | Our own first-class constraint type. The advantage over the regular plutus constraint
---  type is that we get to add whatever we need and we hide away the type variables in existentials.
-data Constraint where
-  PaysScript ::
-    (Pl.ToData (Pl.DatumType a), Show (Pl.DatumType a), Typeable a) =>
-    Pl.TypedValidator a ->
-    [(Pl.DatumType a, Pl.Value)] ->
-    Constraint
+-- Our own first-class constraint types. The advantage over the regular plutus constraint
+-- type is that we get to add whatever we need and we hide away the type variables in existentials.
+
+-- | Constraints which do not specify new transaction outputs
```

Comment:
Massive nit-picking; but maybe we first show the `Constraint` type, explain it divides constraints in two different classes of constraint and respects the ordering of `OutConstraint`, then we define the rest of the types. Seeing the resulting type always makes it easier to understand its component parts.

Reply:
I agree to swap the definitions.